### PR TITLE
Added possibility to relax metadata limit enforcement

### PR DESCRIPTION
**Pull request by hbs**

### CODE REVIEW

**Review by pi-r-p — CHANGES_REQUESTED** [review]
When using UPDATE or META, there is no clear error on WarpScript side (using huge labels or attribute)

```
[11:15:09] ERROR file://Untitled-95:22 Exception at 'DUP $wt =>UPDATE<= $wt META ' in section [TOP] (UPDATE failed to complete successfully (Bad Request))
[11:15:28] ERROR file://Untitled-95:22 Exception at 'DUP $wt UPDATE $wt =>META<= ' in section [TOP] (META failed to complete successfully (Bad Request))
```

The meta logs are crazy long too (the huge labels and attribute are logged):
```
2023-06-15T11:18:25,187 Warp Jetty Thread-71 WARN  handler.ErrorHandler - Error page too large: 400 Invalid metadata Metadata(name:crazylongmetadata, labels:{.app=demo.CHANGEME, .producer=f703eb19-ff99-4fe9-bd44-62b2dd000b14, .owner=f703eb19-ff99-4fe9-bd44-62b2dd000b14}, classId:0, labelsId:0, attributes:{hugeattribute=UEsDBBQAAAAAAOmLz1ANvtUaAQAAAAEAAAAHAAAAdmVyc2lvbjJQSwMEFAAAAAgA6YvPUDsWe/ePAAAAyAAAAAgAAABtZXRhZGF0YS3LOw6DMBBF0X5W4Q0EYf7NdFOkyQoQxUAmyIqDLdtQsPpEMd2R7nvjat3MdoJo1uDe6pAQjduwLNqiBBifcphFlJ5gYZ/2IC9jBa1bzXLTkFxiq3xws0QcIPLHWwmcBAf1uJ9X541/ByzhP9RIlyoknVUjVVkNUp3VIjVZHVKb1SN1WQNSD/tmUjSnoIYvUEsDBBQAAAAIAOmLz1C5BOaRzycAAAAAoAAJAAAAbG9naWMtMS0x7MGBAAAAAIAgf94FFqkKAAAA( ... 6 MB more)
```

There is no update error in the log. So updating huge metadata fails without error.

**Review by pi-r-p — CHANGES_REQUESTED** [review]
Setting limits sky high leads to a server error whatever the length of metadata.
There is no explanation in the logs nor in the WarpScript error

```
warp.class.maxsize = 10240000
warp.labels.maxsize = 20480000
warp.attributes.maxsize = 81920000
```


```
Exception at 'DUP $wt =>UPDATE<= $wt META ' in section [TOP] (UPDATE failed to complete successfully (Server Error))
```

With the same limit and `warp.relax.metadata.maxsize = true` , it works with the warning in the logs.

File: warp10/src/main/java/io/warp10/continuum/MetadataUtils.java [hunk 1]
```diff
@@ -67,10 +72,23 @@ public class MetadataUtils {
...
+    // This is checked for all versions of Warp 10 since they could either store directly
+    // in FoundationDB (standalone+ / distributed) or have a datalog replication in place
+    // to another instance which could be using FoundationDB.
+    // The check can be relaxed in the case of a standalone version by setting the config
+    // key 'warp.relax.metadata.limit' to 'true'
     //
 
     if (MAX_CLASS_SIZE + MAX_LABELS_SIZE + MAX_ATTRIBUTES_SIZE > FDBUtils.MAX_VALUE_SIZE - FDB_SIZE_RESERVED) {
-      throw new RuntimeException("Invalid metadata limits, '" + Configuration.WARP_CLASS_MAXSIZE + "' + '" + Configuration.WARP_LABELS_MAXSIZE + "' + '" + Configuration.WARP_ATTRIBUTES_MAXSIZE + "' should be less than " + (FDBUtils.MAX_VALUE_SIZE - FDB_SIZE_RESERVED));
+      if (!WarpConfig.isStandaloneMode() || Constants.BACKEND_FDB.equals(WarpConfig.getProperty(Configuration.BACKEND)) || !"true".equals(WarpConfig.getProperty(Configuration.WARP_RELAX_METADATA_MAXSIZE, "false"))) {
```
Comment (pi-r-p) [inline warp10/src/main/java/io/warp10/continuum/MetadataUtils.java:83]: condition problem, when limits are higher, it fails with `Invalid metadata limits, 'warp.class.maxsize' + 'warp.labels.maxsize' + 'warp.attributes.maxsize' should be less than 98976 to ensure compatibility with FoundationDB based instances` message whatever relax configuration value is.

**Review by pi-r-p — CHANGES_REQUESTED** [review]

File: warp10/src/main/java/io/warp10/continuum/MetadataUtils.java [hunk 2]
```diff
@@ -67,10 +72,23 @@ public class MetadataUtils {
...
     //
     // Ensure that the limits will allow for serialized metadata to fit in FoundationDB
+    // This is checked for all versions of Warp 10 since they could either store directly
+    // in FoundationDB (standalone+ / distributed) or have a datalog replication in place
+    // to another instance which could be using FoundationDB.
+    // The check can be relaxed in the case of a standalone version by setting the config
+    // key 'warp.relax.metadata.limit' to 'true'
     //
 
     if (MAX_CLASS_SIZE + MAX_LABELS_SIZE + MAX_ATTRIBUTES_SIZE > FDBUtils.MAX_VALUE_SIZE - FDB_SIZE_RESERVED) {
```
Comment (pi-r-p) [inline warp10/src/main/java/io/warp10/continuum/MetadataUtils.java:82]: This can overflow.

**Review by pi-r-p — CHANGES_REQUESTED** [review]

File: etc/conf.templates/standalone/00-warp.conf.template [hunk 3]
```diff
@@ -252,6 +257,13 @@ warp.token.read.attributes.default = { '.cap:limits' '' }
 //
 #warp.attributes.maxsize = 8192
 
+//
+// When using a standalone version with no replication to a FoundationDB based instance
+// the limit of metadata size (class + attribtues + labels) can be relaxed by setting this
+// configuration to 'true' to simply emit a warning.
+//
+#warp.relax.metadata.maxsize = true
+
```
Comment (pi-r-p) [inline etc/conf.templates/standalone/00-warp.conf.template:266]: Typo on attributes, and may be rephrased:

_By default, metadata size is limited to the maximum allowed by fdb.
When using a leveldb standalone version, the limit of metadata size (class + attributes + labels) can be relaxed by setting this configuration to 'true' to simply emit a warning.
This parameter has no effect on standalone+ (fdb backend).
Relaxing the limit means that data cannot be copied to an fdb backend in the future._

**Review by pi-r-p — CHANGES_REQUESTED** [review]

**Review by pi-r-p — APPROVED** [review]

**Review by steven-gueguen — APPROVED** [review]

**Pull request merged**